main/logger.c

```diff
@@ -1645,6 +1646,266 @@ static char *handle_logger_remove_channel(struct ast_cli_entry *e, int cmd, stru
+{
+	char **callid_group;
+	callid_group = ast_threadstorage_get(&callid_group_name, sizeof(*callid_group));
+	return callid_group ? *callid_group : NULL;
+}
+
+static int ast_callid_set_chanloggroup(const char *group)
+{
+	/* Use threadstorage for constant time access, rather than a linked list */
+	ast_callid callid;
+	char **callid_group;
+
+	callid = ast_read_threadstorage_callid();
+	if (!callid) {
+		/* Should never be called on non-PBX threads */
+		ast_assert(0);
+		return -1;
+	}
+
+	callid_group = ast_threadstorage_get(&callid_group_name, sizeof(*callid_group));
+
+	if (!group) {
+		/* Remove from list */
+		if (!*callid_group) {
+			return 0; /* Wasn't in any group to begin with */
+		}
+		ast_free(*callid_group);
+		return 0; /* Set Call ID group for the first time */
+	}
+	/* Existing group */
+	ast_free(*callid_group);
+	*callid_group = ast_strdup(group);
+	if (!*callid_group) {
+		return -1;
+	}
+	return 0; /* Set Call ID group for the first time */
+}
+
+static int callid_group_remove_filters(void)
+{
+	int i = 0;
+	struct chan_group_lock *cgl;
+
+	AST_RWLIST_WRLOCK(&chan_group_lock_list);
+	while ((cgl = AST_RWLIST_REMOVE_HEAD(&chan_group_lock_list, entry))) {
+		ast_free(cgl);
+		i++;
+	}
+	callid_filtering = 0;
+	AST_RWLIST_UNLOCK(&chan_group_lock_list);
+	return i;
+}
+
+static int callid_group_set_filter(const char *group, int enabled)
+{
+	struct chan_group_lock *cgl;
+
+	AST_RWLIST_WRLOCK(&chan_group_lock_list);
+	AST_RWLIST_TRAVERSE_SAFE_BEGIN(&chan_group_lock_list, cgl, entry) {
+		if (!strcmp(group, cgl->name)) {
+			if (!enabled) {
+				AST_RWLIST_REMOVE_CURRENT(entry);
+				ast_free(cgl);
+			}
+			break;
+		}
+	}
+	AST_RWLIST_TRAVERSE_SAFE_END;
+
+	if (!enabled) {
+		if (AST_LIST_EMPTY(&chan_group_lock_list)) {
+			callid_filtering = 0;
+		}
+		AST_RWLIST_UNLOCK(&chan_group_lock_list);
+		return 0;
+	}
+
+	if (!cgl) {
+		cgl = ast_calloc(1, sizeof(*cgl) + strlen(group) + 1);
+		if (!cgl) {
+			AST_RWLIST_UNLOCK(&chan_group_lock_list);
+			return -1;
+		}
+		strcpy(cgl->name, group); /* Safe */
+		AST_RWLIST_INSERT_HEAD(&chan_group_lock_list, cgl, entry);
+	} /* else, already existed, and was already enabled, no change */
+	callid_filtering = 1;
+	AST_RWLIST_UNLOCK(&chan_group_lock_list);
+	return 0;
+}
+
+static int callid_logging_enabled(void)
+{
+	struct chan_group_lock *cgl;
+	const char *callidgroup;
+
+	if (!callid_filtering) {
+		return 1; /* Everything enabled by default, if no filtering */
+	}
+
+	callidgroup = ast_get_callid_group();
+	if (!callidgroup) {
+		return 0; /* Filtering, but no call group, not enabled */
+	}
+
+	AST_RWLIST_RDLOCK(&chan_group_lock_list);
+	AST_RWLIST_TRAVERSE(&chan_group_lock_list, cgl, entry) {
+		if (!strcmp(callidgroup, cgl->name)) {
+			break;
+		}
+	}
+	AST_RWLIST_UNLOCK(&chan_group_lock_list);
+	return cgl ? 1 : 0; /* If found, enabled, otherwise not */
+}
+
+/*** DOCUMENTATION
+	<application name="SetChannelLogGroup" language="en_US">
```

Comment:
I don't think this should be a dialplan application. It should be a dialplan function to set and get, which allows it to be set in a PJSIP endpoint definition for example, or when originated in AMI.